kyo-core/shared/src/main/scala/kyo/Maybe.scala

```diff
@@ -0,0 +1,159 @@
+    given [T]: Conversion[Maybe[T], IterableOnce[T]]                 = _.iterator
+    given [T: Flat]: Flat[Maybe[T]]                                  = Flat.unsafe.bypass
+
+    def apply[T](v: T): Maybe[T] =
+        if isNull(v) then
+            Empty
+        else
+            v match
+                case v: DefinedEmpty => v.nest
+                case v               => v
+
+    def fromScala[T](opt: Option[T]): Maybe[T] =
+        opt match
+            case Some(v) => Defined(v)
+            case None    => Empty
+
+    def empty[T]: Maybe[T] = Empty
+
+    def when[T](cond: Boolean)(v: => T): Maybe[T] =
+        if cond then v else Empty
+
+    opaque type Defined[+T] = T | DefinedEmpty
+
+    object Defined:
+
+        def apply[T](v: T): Defined[T] =
+            v match
+                case v: DefinedEmpty => v.nest
+                case v: Empty        => DefinedEmpty()
+                case v               => v
+
+        // TODO avoid allocation
+        def unapply[T](opt: Maybe[T]): Option[T] =
+            opt.toScala
+    end Defined
+
+    sealed abstract class Empty
+    case object Empty extends Empty
+
+    extension [T](self: Maybe[T])
+
+        inline def toScala: Option[T] =
+            if isEmpty then None
+            else Some(get)
+
+        inline def isEmpty: Boolean =
+            self match
+                case _: Empty => true
+                case _        => false
+
+        inline def isDefined: Boolean = !isEmpty
+
+        inline def nonEmpty: Boolean = !isEmpty
+
+        // TODO compilation failure if inlined
+        def get: T =
+            (self: @unchecked) match
+                case _: Empty =>
+                    throw new NoSuchElementException("Maybe.get")
+                case self: DefinedEmpty =>
+                    self.unnest.asInstanceOf[T]
+                case v: T @unchecked =>
+                    v
+
+        inline def getOrElse[B >: T](inline default: => B): B =
+            if isEmpty then default else get
+
+        inline def fold[U](inline ifEmpty: => U, inline ifDefined: T => U): U =
+            if isEmpty then ifEmpty else ifDefined(get)
+
+        inline def map[U](inline f: T => U): Maybe[U] =
+            if isEmpty then Empty else f(get)
+
+        inline def flatMap[U](inline f: T => Maybe[U]): Maybe[U] =
+            if isEmpty then Maybe.empty else f(get)
+
+        inline def flatten[U](using inline ev: T <:< Maybe[U]): Maybe[U] =
+            if isEmpty then Empty else ev(get)
+
+        inline def filter(inline f: T => Boolean): Maybe[T] =
+            if isEmpty || f(get) then self else Empty
+
+        inline def filterNot(inline f: T => Boolean): Maybe[T] =
+            if isEmpty || !f(get) then self else Empty
+
+        // TODO compilation failure if inlined
+        def contains[U](elem: U)(using CanEqual[T, U]): Boolean =
+            !isEmpty && self.get == elem
```

Comment:
Does compilation fail if you remove the `self.` prefix from `get`?

Reply:
It does. I've removed `self.`. I had added it because the compiler has trouble when I inline this method.